hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java

```diff
@@ -736,23 +744,48 @@ public void notifyTermIndexUpdated(long term, long index) {
     return f;
   }
 
+  // Removes the stateMachine data from cache once both followers catch up
+  // to the particular index.
+  private void removeStateMachineDataIfNeeded(long index) {
+    if (waitOnBothFollowers) {
+      LOG.info("Removing data corresponding to log index {} from cache",
+          index);
+      try {
+        RaftServer.Division division = ratisServer.getServer().getDivision(gid);
+        if (division.getInfo().isLeader() && Arrays
+            .stream(division.getInfo().getFollowerNextIndices())
+            .allMatch(i -> i >= index)) {
+          LOG.debug("Removing data corresponding to log index {} from cache",
+              index);
+          stateMachineDataCache.remove(index);
```

Comment:
We might not remove some entries in case the condition check returns false.

Reply:
I think, the condition will be false only, if a follower is not able to catch up . In this case, the pipeline will be closed down anyways. Other situation can be case, where there are few entries remaining, there are no more requests coming in, no more log transactions, no more configuration or metadata transactions being done and the pipeline is just idle. In such cases, i beleive its ok to have the entries in the cache. If a new request comes in, entries will be removed .

I think We don't need to handle this particular corner case.

Reply:
`Arrays
            .stream(division.getInfo().getFollowerNextIndices())
            .allMatch(i -> i >= index)`
This condition can be false for some indices. For example `removeStateMachineDataIfNeeded` is called for every index. It is possible that next index could be lagging for one follower. As a result condition eviction would happen for some indexes and not for others.

Reply:
I think, this is the intention right. Only when both followers catch up, eviction should happen not for others.

Reply:
But we might end up never evicting that entry. We are not evicting the entry again unless there is a leader change.

Reply:
i see..Got ur point.

Reply:
Adressed in the latest patch.